File: 00_leetcode/java/1846_maximum_element_after_decreasig_and_rearranging.java
/**
 * 1846. Maximum Element After Decreasing and Rearranging
 * Problem Link: https://leetcode.com/problems/maximum-element-after-decreasing-and-rearranging/
*/

class Solution {
    public int maximumElementAfterDecrementingAndRearranging(int[] arr) {
        int n = arr.length;
        int[] count = new int[n + 1];
        
        for (int num : arr) {
            count[Math.min(num, n)]++;
        }

        int max = 0;
        for (int i = 1; i <= n; i++) {
            max = Math.min(max + count[i], i); 
        }

        return max;
    }
}

File: 00_leetcode/python3/1846_maximum_element_after_decreasig_and_rearranging.py
"""
 * 1846. Maximum Element After Decreasing and Rearranging
 * Problem Link: https://leetcode.com/problems/maximum-element-after-decreasing-and-rearranging/
"""

from typing import List

class Solution:
    def maximumElementAfterDecrementingAndRearranging(self, arr: List[int]) -> int:
        arr.sort() 
        max_val = 1 

        for i in range(1, len(arr)):
            max_val = min(max_val + 1, arr[i])  
            
        return max_val